### PR TITLE
support memory mapped tpms 

### PR DESCRIPTION
**Pull request by rizlik**
Enabled with `--enable-mmio` or `WOLFTPM_MMIO`.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -8,7 +8,7 @@ Portable TPM 2.0 project designed for embedded use.
 * This implementation provides all TPM 2.0 API’s in compliance with the specification.
 * Wrappers provided to simplify Key Generation/Loading, RSA encrypt/decrypt, ECC sign/verify, ECDH, NV, Hashing/Hmac and AES.
 * Testing done using the following TPM 2.0 modules: STM ST33TP* SPI/I2C, Infineon OPTIGA SLB9670/SLB9672, Microchip ATTPM20, Nations Tech Z32H330TC and Nuvoton NPCT650/NPCT750.
-* wolfTPM uses the TPM Interface Specification (TIS) to communicate over SPI.
+* wolfTPM uses the TPM Interface Specification (TIS) to communicate over SPI or using a memory mapped io range.
```
Comment (danielinux) [inline README.md:11]: review: uses [...] TIS or [using] a memory mapped? remove "using"?

File: src/tpm2_tis.c [hunk 2]
```diff
@@ -57,8 +57,13 @@ enum tpm_tis_int_flags {
     TPM_INTF_DATA_AVAIL_INT     = 0x001,
 };
 
-
+#ifndef TPM_BASE_ADDRESS
+#ifdef WOLFTPM_MMIO_BUILTIN_CB
```
Comment (danielinux) [inline src/tpm2_tis.c:61]: I think this define could simply be WOLFTPM_MMIO

**Review by danielinux — CHANGES_REQUESTED** [review]
- please review README change.
- Consider shortening the option name to WOLFTPM_MMIO
- Should we also add a configure.ac option like `--enable-mmio` that would activate MMIO access?

File: README.md (hunk 1 above)
Comment (danielinux) [reply README.md:11]: I/O should also be capitalized

File: src/tpm2_tis.c (hunk 2 above)
Comment (rizlik) [reply src/tpm2_tis.c:61]: I think that `WOLFTPM_MMIO_BUILTIN_CB` clarify better that custom I/O callbacks cannot be used together with this feature.

**Review by danielinux — APPROVED** [review]

**Comment by danielinux** [comment]
Good to go, tested on Qemu. Off to @dgarske for final review & configuration naming (see open  comment)

**Review by dgarske — DISMISSED** [review]

**Review by jpbland1 — APPROVED** [review]
tested working on my setup

**Review by dgarske — DISMISSED** [review]

**Review by jpbland1 — DISMISSED** [review]
Compiled on my setup, looks good

File: wolftpm/tpm2_tis.h [hunk 3]
```diff
@@ -54,6 +54,11 @@ WOLFTPM_LOCAL int TPM2_TIS_StartupWait(TPM2_CTX* ctx, int timeout);
 WOLFTPM_LOCAL int TPM2_TIS_Write(TPM2_CTX* ctx, word32 addr, const byte* value, word32 len);
 WOLFTPM_LOCAL int TPM2_TIS_Read(TPM2_CTX* ctx, word32 addr, byte* result, word32 len);
 
+#ifdef WOLFTPM_MMIO_BUILTIN_CB
```
Comment (jpbland1) [inline wolftpm/tpm2_tis.h:57]: this is in the hal file and now has the wrong flag

File: hal/tpm_io_mmio.c [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+/* Included via tpm_io.c if WOLFTPM_INCLUDE_IO_FILE is defined */
+#ifdef WOLFTPM_INCLUDE_IO_FILE
+#ifdef WOLFTPM_MMIO
+
+#ifndef MIMO_BASE_ADDRESS
+#define MIMO_BASE_ADDRESS 0xFE000000u
+#endif
+
+#ifndef WOLFTPM_ADV_IO
+#error "WOLFTPM_MMIO_BUILTIN_CB requires WOLFTPM_ADV_IO"
```
Comment (jpbland1) [inline hal/tpm_io_mmio.c:42]: rename to new flag

**Review by jpbland1 — APPROVED** [review]
Got it working with the qemu repo

**Review by dgarske — APPROVED** [review]

**Pull request merged**